File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque

dx = [0,0,1,-1]
dy = [1,-1,0,0]

n = int(input())

graph = []
result= []

for i in range(n):
    graph.append(list(map(int,input())))
def bfs(graph,a,b):
    q = deque()
    q.append([a,b])
    graph[a][b] = 0
    count = 1

    while q:
        x,y = q.popleft()
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]

            if nx<0 or nx>=n or ny<0 or ny>=n:
                continue
            if graph[nx][ny] == 1:
                graph[nx][ny] = 0
                q.append([nx,ny])
                count += 1
    return count

for i in range(n):
    for j in range(n):
        if(graph[i][j]==1):
            result.append(bfs(graph,i,j))

result.sort()
print(len(result))
for i in result:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;

public class Main {
    static int[][] matrix;
    static int n;
    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};
    static ArrayList<Integer> result;


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] st = br.readLine().split(" ");

        n = Integer.parseInt(st[0]);

        matrix = new int[n][n];
        result = new ArrayList<>();

        for(int i = 0; i < n; i++){
            String inp = br.readLine();

            for(int j = 0; j < n; j++){
                matrix[i][j] = inp.charAt(j) - '0';
            }

        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(matrix[i][j] == 1){
                    result.add(bfs(i, j));
                }
            }
        }
        Collections.sort(result);
        System.out.println(result.size());
        for(int i : result){
            System.out.println(i);
        }
    }
    public static int bfs(int x, int y){
        Queue<Point> q = new LinkedList<>();
        q.add(new Point(x, y));
        matrix[x][y] = 0;
        int count = 1;
        while(!q.isEmpty()){
            Point current = q.poll();

            for(int i = 0; i < 4; i++){
                int nextX = current.x + dx[i];
                int nextY = current.y + dy[i];

                if(nextX < 0 || nextY < 0 || nextX >= n || nextY >= n){
                    continue;
                }

                if(matrix[nextX][nextY] == 1){
                    matrix[nextX][nextY] = 0;
                    q.add(new Point(nextX, nextY));
                    count += 1;
                }
            }
        }
        return count;
    }
}
class Point{
    int x;
    int y;
    Point(int x, int y){
        this.x = x;
        this.y = y;
    }
}
